### Today we are going to perform the simple classification of the amazon reviews' sentiment.

### Please, download the dataset amazon_baby.csv.

In [108]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import string
import sys
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
import time
def remove_punctuation(text):
    import string
    translator = str.maketrans('', '', string.punctuation)
    return text.translate(translator)

baby_df = pd.read_csv('amazon_baby.csv')
baby_df.head()

name  \
0                                                                           Planetwise Flannel Wipes   
1                                                                              Planetwise Wipe Pouch   
2                                                                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumbuddy To Love's Binky Fairy Puppet and Adorable Book   
4  Stop Pacifier Sucking without tears with Thumbuddy To Love's Binky Fairy Puppet and Adorable Book   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                 review  \
0  These flannel wipes are OK, but in my opinion not worth keeping.  I also ordered someImse Vimse Cloth Wipes-Ocean Blue-12 countwhich are larger, had a nicer, softer texture and just seemed higher quality.  I use cloth wipes for hands and faces and have been usingThirsties 6 Pack Fab Wipes, Boyfor about 8 months now and need to replace them because they are starting to get rough and have had stink issues for a while that stripping no longer handles.   
1                                                                                                                                                                                                                                                                                                        it came early and was not disappointed. i love planet wise bags and now my wipe holder. it keps my osocozy wipes moist and does not leak. highly recommend it.   
2                                                                                                                                                                                                                                                                                                                       Very soft and comfortable and warmer than it looks...fit the full size bed perfectly...would recommend to anyone looking for this type of quilt   
3                                                                This is a product well worth the purchase.  I have not found anything else like this, and it is a positive, ingenious approach to losing the binky.  What I love most about this product is how much ownership my daughter has in getting rid of the binky.  She is so proud of herself, and loves her little fairy.  I love the artwork, the chart in the back, and the clever approach of this tool.   
4                                                  All of my kids have cried non-stop when I tried to ween them off their pacifier, until I found Thumbuddy To Love's Binky Fairy Puppet.  It is an easy way to work with your kids to allow them to understand where their pacifier is going and help them part from it.This is a must buy book, and a great gift for expecting parents!!  You will save them soo many headaches.Thanks for this book!  You all rock!!   

   rating  
0       3  
1       5  
2       5  
3       5  
4       5

## Exercise 1 (data preparation)
a) Remove punctuation from reviews using the given function.   
b) Replace all missing (nan) revies with empty "" string.  
c) Drop all the entries with rating = 3, as they have neutral sentiment.   
d) Set all positive ($\geq$4) ratings to 1 and negative($\leq$2) to -1.

In [109]:
#b)
baby_df['review'] = baby_df['review'].fillna("")

#short test:
baby_df["review"][38] == baby_df["review"][38]

True

In [110]:
#a)
baby_df['review'] = baby_df['review'].apply(remove_punctuation)

#short test: 
baby_df["review"][4] == 'All of my kids have cried nonstop when I tried to ween them off their pacifier until I found Thumbuddy To Loves Binky Fairy Puppet  It is an easy way to work with your kids to allow them to understand where their pacifier is going and help them part from itThis is a must buy book and a great gift for expecting parents  You will save them soo many headachesThanks for this book  You all rock'
remove_punctuation(baby_df["review"][4]) == 'All of my kids have cried nonstop when I tried to ween them off their pacifier until I found Thumbuddy To Loves Binky Fairy Puppet  It is an easy way to work with your kids to allow them to understand where their pacifier is going and help them part from itThis is a must buy book and a great gift for expecting parents  You will save them soo many headachesThanks for this book  You all rock'

True

In [111]:
#c)
baby_df = baby_df[baby_df['rating'] != 3]
#short test:
sum(baby_df["rating"] == 3)

0

In [112]:
#d) 
baby_df["rating"]=baby_df["rating"].apply(lambda x:1 if x>=4 else -1)
#short test:
sum(baby_df["rating"]**2 != 1)

0

In [113]:
baby_df.head(30)

name  \
1                                                                               Planetwise Wipe Pouch   
2                                                                 Annas Dream Full Quilt with 2 Shams   
3   Stop Pacifier Sucking without tears with Thumbuddy To Love's Binky Fairy Puppet and Adorable Book   
4   Stop Pacifier Sucking without tears with Thumbuddy To Love's Binky Fairy Puppet and Adorable Book   
5   Stop Pacifier Sucking without tears with Thumbuddy To Love's Binky Fairy Puppet and Adorable Book   
6                                                             A Tale of Baby's Days with Peter Rabbit   
7                                           Baby Tracker&reg; - Daily Childcare Journal, Schedule Log   
8                                           Baby Tracker&reg; - Daily Childcare Journal, Schedule Log   
9                                           Baby Tracker&reg; - Daily Childcare Journal, Schedule Log   
10                                          Baby Tracker&reg; - Daily Childcare Journal, Schedule Log   
11                                          Baby Tracker&reg; - Daily Childcare Journal, Schedule Log   
12                                          Baby Tracker&reg; - Daily Childcare Journal, Schedule Log   
14                                                     Nature's Lullabies First Year Sticker Calendar   
15                                                     Nature's Lullabies First Year Sticker Calendar   
16                                                     Nature's Lullabies First Year Sticker Calendar   
17                                                    Nature's Lullabies Second Year Sticker Calendar   
18                                                    Nature's Lullabies Second Year Sticker Calendar   
19                                                    Nature's Lullabies Second Year Sticker Calendar   
20                                                    Nature's Lullabies Second Year Sticker Calendar   
21                                                    Nature's Lullabies Second Year Sticker Calendar   
22                                                    Nature's Lullabies Second Year Sticker Calendar   
24                                                    Nature's Lullabies Second Year Sticker Calendar   
25                                                    Nature's Lullabies Second Year Sticker Calendar   
26                                                                       Baby's First Journal - Green   
28                                                                        Lamaze Peekaboo, I Love You   
29                                                                        Lamaze Peekaboo, I Love You   
30                                                                        Lamaze Peekaboo, I Love You   
31                                                                        Lamaze Peekaboo, I Love You   
32                                                                        Lamaze Peekaboo, I Love You   
33                                                                        Lamaze Peekaboo, I Love You   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         

## Comment: 
First, I replaced any missing values with empty strings to avoid issues with later transformations. Then, I applied a function to each review to remove punctuation. After that, I filtered out all entries with a rating of 3, as they represent neutral sentiment, which we’re not including in this dataset. Finally, I assigned a value of 1 to all positive ratings (4 or 5) and -1 to all negative ratings (1 or 2).

## CountVectorizer
In order to analyze strings, we need to assign them numerical values. We will use one of the simplest string representation, which transforms strings into the $n$ dimensional vectors. The number of dimensions will be the size of our dictionary, and then the values of the vector will represent the number of appereances of the given word in the sentence.

In [114]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
reviews_train_example = ["We like apples",
                   "We hate oranges",
                   "I adore bananas",
                   "We like like apples and oranges",
                   "They dislike bananas"]

X_train_example = vectorizer.fit_transform(reviews_train_example)

print(vectorizer.get_feature_names())
print(X_train_example.todense())



['adore', 'and', 'apples', 'bananas', 'dislike', 'hate', 'like', 'oranges', 'they', 'we']
[[0 0 1 0 0 0 1 0 0 1]
 [0 0 0 0 0 1 0 1 0 1]
 [1 0 0 1 0 0 0 0 0 0]
 [0 1 1 0 0 0 2 1 0 1]
 [0 0 0 1 1 0 0 0 1 0]]


D:\Anaconda\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In [115]:
reviews_test_example = ["They like bananas",
                   "We hate oranges bananas and apples",
                   "We love bananas"] #New word!

X_test_example = vectorizer.transform(reviews_test_example)

print(X_test_example.todense())

[[0 0 0 1 0 0 1 0 1 0]
 [0 1 1 1 0 1 0 1 0 1]
 [0 0 0 1 0 0 0 0 0 1]]


We should acknowledge few facts. Firstly, CountVectorizer does not take order into account. Secondly, it ignores one-letter words (this can be changed during initialization). Finally, for test values, CountVectorizer ignores words which are not in it's dictionary.

## Exercise 2 
a) Split dataset into training and test sets.     
b) Transform reviews into vectors using CountVectorizer. 

In [116]:
#a)
X=baby_df["review"]
Y=baby_df["rating"]
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)


In [117]:
#b)
vectorizer = CountVectorizer()
X_train_vector = vectorizer.fit_transform(X_train)
X_test_vector = vectorizer.transform(X_test)
print(vectorizer.get_feature_names_out())

['00' '000' '0001' ... 'zzzzzz' 'zzzzzzz' 'zzzzzzzzzzz']


## Comment
I split the dataset into training and testing sets, using 80% for training and 20% for testing. Then, I used CountVectorizer to transform the review text into numerical feature vectors, which represent the occurrence of words in the reviews. The vectorizer was fitted on the training data, and the test data was transformed using the fitted model. Finally, I retrieved the list of unique words from the training set, which were used as features in the model. What is more, I noticed that some of the features listed are not strictly words, which can lead to issues when building the model.


## Exercise 3 
a) Train LogisticRegression model on training data (reviews processed with CountVectorizer, ratings as they were).   
b) Print 10 most positive and 10 most negative words.

In [118]:
#a)
model = LogisticRegression(solver='sag',max_iter=1000)
model.fit(X_train_vector, Y_train)


D:\Anaconda\anaconda3\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


LogisticRegression(max_iter=1000, solver='sag')

In [119]:
#b)
# Get feature names (words) from the vectorizer
feature_names = np.array(vectorizer.get_feature_names_out())

# Get the coefficients of the trained model
coefficients = model.coef_.flatten()

# Sort the indices of the coefficients
sorted_indices = coefficients.argsort()

# Get the 10 most negative and 10 most positive words
most_negative_words = feature_names[sorted_indices[:10]]
most_positive_words = feature_names[sorted_indices[-10:]]

print("Most Negative Words:", most_negative_words)
print("Most Positive Words:", most_positive_words)
#hint: model.coef_, vectorizer.get_feature_names()

Most Negative Words: ['useless' 'disappointed' 'worst' 'disappointing' 'returned' 'poor'
 'poorly' 'waste' 'terrible' 'returning']
Most Positive Words: ['highly' 'exactly' 'pleased' 'glad' 'perfectly' 'love' 'awesome' 'loves'
 'excellent' 'perfect']


## Comment
To predict ratings, I started by training a logistic regression model on the training data. 

Given the complexity of the dataset, I opted for a more efficient solver - SAG (Stochastic Average Gradient) , which is better suited for larger datasets ( We have more than 166000 records) . Additionally, I adjusted the maximum number of iterations to ensure the model had enough time to converge.

By analyzing the feature importance in the decision function, I was able to identify the most positive and negative words. Before using the SAG solver, the results for positive words seemed a bit unusual (e.g., "ply"). However, with the SAG solver, words related to positive human emotions were more frequently associated with positive reviews, while the word "return" emerged as the most common in negative reviews. This demonstrates how sentiment in text is often closely tied to specific keywords.

## Exercise 4 
a) Predict the sentiment of test data reviews.   
b) Predict the sentiment of test data reviews in terms of probability.   
c) Find five most positive and most negative reviews.   
d) Calculate the accuracy of predictions.

In [120]:
#a)
Y_pred = model.predict(X_test_vector)


In [121]:
#b)
Y_pred_prob = model.predict_proba(X_test_vector)

print("Probabilities for test data:", Y_pred_prob[:10])

Probabilities for test data: [[0.42565238 0.57434762]
 [0.72584614 0.27415386]
 [0.61486256 0.38513744]
 [0.00363607 0.99636393]
 [0.00798456 0.99201544]
 [0.00267802 0.99732198]
 [0.01876934 0.98123066]
 [0.15202413 0.84797587]
 [0.03499297 0.96500703]
 [0.84356014 0.15643986]]


## Comment
In part (a), I trained a logistic regression model and used it to predict the sentiment of the test data reviews. The model's predictions were made by assigning each review a sentiment label (positive or negative). In part (b), I predicted the sentiment of the test data reviews in terms of probability, using the predict_proba() method. This method provides the probabilities of a review belonging to each class (positive or negative), which allows us to understand the model's confidence in its predictions. The output contains two probabilities for each review: one for the negative class and one for the positive class.

In [122]:
# Get the index of the highest and lowest probabilities for positive sentiment
positive_indices = np.argsort(Y_pred_prob[:, 1])[-5:]  # Top 5 most positive
negative_indices = np.argsort(Y_pred_prob[:, 1])[:5]  # Top 5 most negative

# Get the corresponding reviews and their IDs
positive_reviews = X_test.iloc[positive_indices]
negative_reviews = X_test.iloc[negative_indices]

# Display the reviews with their IDs
print("Most Positive Reviews:")
for idx in positive_indices:
    print(f"ID: {X_test.index[idx]}, Review: {X_test.iloc[idx][:100]}...")  # Display first 100 characters

print("\nMost Negative Reviews:")
for idx in negative_indices:
    print(f"ID: {X_test.index[idx]}, Review: {X_test.iloc[idx][:100]}...")  # Display first 100 characters


Most Positive Reviews:
ID: 100166, Review: I bought this carrier when my daughter was about 4 weeks old shes now 10 weeks old  I had a Moby tha...
ID: 123632, Review: I did a TON of research before I purchased this stroller and I still watch stroller reviews all the ...
ID: 168086, Review: Buttons vs Best Bottoms reviewFirst thing I want to mention  Best Bottoms inserts are compatible wit...
ID: 129722, Review: This is a review of the 2012 Bumbleride Flite in Ruby color Our baby is 4 weeks now and we got the s...
ID: 50735, Review: The joovy zoom 360 was the perfect solution for us We couldnt justify spending the money on a mounta...

Most Negative Reviews:
ID: 147902, Review: My disappointment with this product prompted me to write to the manufacturer  Below is my correspond...
ID: 175191, Review: I had to return this stroller for three reasons1 Straps are too small and my 15 month old who is ave...
ID: 89902, Review: I am so incredibly disappointed with the stroller I just dont even 

## Comment
Here, we have 10 product reviews, consisting of 5 most positive and 5 most negative reviews. After reading these reviews, it is evident that the model classified them accurately, as my own judgment aligns with the model's predictions. This suggests that the model is well-trained and capable of identifying sentiment effectively. It is a positive indicator of the model's performance, reflecting its ability to generalize to real-world data.

In [123]:
#d) 
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(Y_test, Y_pred)
print(f"Model accuracy: {round(accuracy * 100, 2)}%")

Model accuracy: 93.57%


## Comment
he model achieved an accuracy of 93.57%, which indicates that it correctly predicted the sentiment of the reviews in over 93% of the cases. This high accuracy suggests that the model performs well in classifying the sentiment of product reviews, demonstrating its effectiveness in this task.


## Exercise 5
In this exercise we will limit the dictionary of CountVectorizer to the set of significant words, defined below.


a) Redo exercises 2-5 using limited dictionary.   
b) Check the impact of all the words from the dictionary.   
c) Compare accuracy of predictions and the time of evaluation.

In [124]:
significant_words = ['love','great','easy','old','little','perfect','loves','well','able','car','broke','less','even','waste','disappointed','work','product','money','would','return']


In [128]:
#a)
X = baby_df["review"]
Y = baby_df["rating"]
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

# Use the limited dictionary in CountVectorizer
vectorizer = CountVectorizer(vocabulary=significant_words)
X_train_vector_lim = vectorizer.fit_transform(X_train)
X_test_vector_lim = vectorizer.transform(X_test)

# Train Logistic Regression model
model = LogisticRegression(solver='sag' ,max_iter=1000)
model.fit(X_train_vector_lim, y_train)

# Make predictions
y_pred = model.predict(X_test_vector_lim)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)

print(f"Model accuracy with limited dictionary: {round(accuracy * 100, 2)}%")


Model accuracy with limited dictionary: 86.9%


## Comment
By limiting the dictionary to a set of significant words, we trained the model in the same way as the previous one. The accuracy achieved was 87%, which, while still decent, is notably lower than the previous score. This highlights how reducing the number of words in the dictionary can negatively affect the model's performance. This experiment demonstrates the importance of selecting an appropriate feature set to optimize model behavior and accuracy.

In [126]:
#b)
coefficients = model.coef_[0]

# Pair each word with its coefficient
word_coefficients = list(zip(vectorizer.get_feature_names_out(), coefficients))

# Sort words by the magnitude of their coefficients (highest impact first)
sorted_words = sorted(word_coefficients, key=lambda x: abs(x[1]), reverse=True)

# Print top 10 most important words
print("Most impactful words and their coefficients:")
for word, coef in sorted_words[:20]:
    print(f"{word}: {coef}")

Most impactful words and their coefficients:
disappointed: -2.388470160730703
return: -2.0773260749529494
waste: -2.0073706299297496
loves: 1.7011900906817243
broke: -1.6668354030521377
perfect: 1.5219676211936388
love: 1.3573491549547172
easy: 1.1916609883788067
money: -0.9385485609847712
great: 0.9311881215715203
work: -0.637371809263575
little: 0.5021118494947115
well: 0.49693814724600927
even: -0.4906573369618201
would: -0.3394232799118203
product: -0.31207322753839395
less: -0.20695811745094433
able: 0.19444741934563314
car: 0.07455446649520711
old: 0.07249836605578674


## Comment
The most impactful words and their coefficients reveal how certain terms affect sentiment classification. Negative words like "disappointed" and "return" are linked to negative sentiment, while positive words like "loves" and "perfect" are associated with positive sentiment. Words with strong positive or negative coefficients indicate their importance in determining the sentiment of the review. This helps us understand which terms the model prioritizes for classification.


In [127]:
#c)
import time

# Full Dictionary Model (Training and Prediction Time)
start_time = time.time()
model_full = LogisticRegression(solver='sag',max_iter=1000).fit(X_train_vector, Y_train)  # Training model with full dictionary
training_time_full = time.time() - start_time

start_time = time.time()
model_full.predict(X_test_vector)  # Prediction with full dictionary model
prediction_time_full = time.time() - start_time

print(f"Full Dictionary Model Training Time: {training_time_full:.4f} seconds")
print(f"Full Dictionary Model Prediction Time: {prediction_time_full:.4f} seconds")
accuracy = accuracy_score(Y_test, Y_pred)
print(f"Model accuracy: {round(accuracy * 100, 2)}%")
# ------------------

# Limited Dictionary Model (Training and Prediction Time)
start_time = time.time()
model_limited = LogisticRegression(solver='sag',max_iter=1000).fit(X_train_vector_lim, y_train)  # Training model with limited dictionary
training_time_limited = time.time() - start_time

start_time = time.time()
model_limited.predict(X_test_vector_lim)  # Prediction with limited dictionary model
prediction_time_limited = time.time() - start_time

print(f"Limited Dictionary Model Training Time: {training_time_limited:.4f} seconds")
print(f"Limited Dictionary Model Prediction Time: {prediction_time_limited:.4f} seconds")
accuracy = accuracy_score(y_test, y_pred)
print(f"Model accuracy: {round(accuracy * 100, 2)}%")

D:\Anaconda\anaconda3\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


Full Dictionary Model Training Time: 95.2294 seconds
Full Dictionary Model Prediction Time: 0.0065 seconds
Model accuracy: 93.57%
Limited Dictionary Model Training Time: 0.3521 seconds
Limited Dictionary Model Prediction Time: 0.0000 seconds
Model accuracy: 86.9%


## Comment

The comparison between the full dictionary and limited dictionary models reveals differences in both performance and accuracy. The full dictionary model takes longer to train and predict due to the larger number of features, which allows it to capture more information from the dataset. This results in higher accuracy, as the model has more words to identify sentiment. In contrast, the limited dictionary model is faster since it works with a smaller set of significant words, but this leads to a decrease in accuracy. The reduced vocabulary does not capture all the nuances of the text, making it less precise in sentiment predictions. While the full dictionary model is slower, it provides better performance in terms of accuracy. Ultimately, the decision between the two models depends on the trade-off between time efficiency and model accuracy for the given application. 
